### PR TITLE
[routing-manager] simplify RA message parsing and preparation

### PR DESCRIPTION
**Pull request by abtink**
This commit adds a new class `RouterAdvertMessage` representing an RA
message. It provides methods to append options (PIO or RIO) to the RA
message or parse and iterate over the options in the RA message
(using newly added `Option::Iterator` which allows use of range-based
`for` loop). These method are used by `RoutingManager` and help
simplify the code.

----

~This PR contains the  commit from~
~- https://github.com/openthread/openthread/pull/7780~

~Please review and check the last commit. Thanks.~

### CODE REVIEW

File: src/core/border_router/routing_manager.cpp [hunk 1]
```diff
@@ -828,128 +828,76 @@ Error RoutingManager::SendRouterSolicitation(void)
...
-            // Set zero route lifetime to immediately invalidate the advertised OMR prefix.
-            rio->Init();
-            rio->SetRouteLifetime(0);
-            rio->SetPrefix(advertisedOmrPrefix.GetPrefix());
-
-            bufferLength += rio->GetSize();
-
-            LogInfo("Stop advertising OMR prefix %s on %s", advertisedOmrPrefix.ToString().AsCString(),
-                    mInfraIf.ToString().AsCString());
+            LogInfo("RouterAdvert: Added RIO for %s (lifetime:0)", omrPrefix.ToString().AsCString());
```
Comment (wgtdkp) [inline src/core/border_router/routing_manager.cpp:871]: nit:
```suggestion
            LogInfo("RouterAdvert: Added RIO for %s (lifetime=0)", omrPrefix.ToString().AsCString());
```

We are using `=` in PIO logs above,

File: src/core/border_router/routing_manager.cpp [hunk 2]
```diff
@@ -828,128 +828,76 @@ Error RoutingManager::SendRouterSolicitation(void)
...
 
-        rio->Init();
-        rio->SetRouteLifetime(kDefaultOmrPrefixLifetime);
-        rio->SetPreference(newOmrPrefix.GetPreference());
-        rio->SetPrefix(newOmrPrefix.GetPrefix());
-
-        bufferLength += rio->GetSize();
-
-        LogInfo("Send OMR prefix %s in RIO (valid lifetime = %u seconds)", newOmrPrefix.ToString().AsCString(),
+        LogInfo("RouterAdvert: Added RIO for %s (lifetime:%u)", omrPrefix.ToString().AsCString(),
```
Comment (wgtdkp) [inline src/core/border_router/routing_manager.cpp:880]: ditto.

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: src/core/border_router/routing_manager.cpp (hunk 1 above)
Comment (abtink) [reply src/core/border_router/routing_manager.cpp]: Changed in new push. Thanks.

File: src/core/net/nd6.hpp [hunk 3]
```diff
@@ -445,101 +455,219 @@ class RouteInfoOption : public Option, private Clearable<RouteInfoOption>
...
      */
-    uint16_t GetRouterLifetime(void) const { return HostSwap16(mRouterLifetime); }
+    Error AppendPrefixInfoOption(const Prefix &aPrefix, uint32_t aValidLifetime, uint32_t aPreferredLifetime);
 
     /**
-     * This method sets the default router preference.
+     * This method appends a Route Info Option to the RA message.
+     *
+     * @param[in] aPrefix             The prefix.
+     * @param[in] aRouterLifetime     The router lifetime in seconds.
```
Comment (jwhui) [inline src/core/net/nd6.hpp:639]:
```suggestion
     * @param[in] aRouteLifetime      The route lifetime in seconds.
```

File: src/core/net/nd6.hpp [hunk 4]
```diff
@@ -445,101 +455,219 @@ class RouteInfoOption : public Option, private Clearable<RouteInfoOption>
...
+     * @param[in] aPreference         The route preference.
      *
-     * @param[in]  aPreference  The router preference.
+     * @retval kErrorNone    Option is appended successfully.
+     * @retval kErrorNoBufs  No more space in the buffer to append the option.
      *
      */
-    void SetDefaultRouterPreference(RoutePreference aPreference);
+    Error AppendRouteInfoOption(const Prefix &aPrefix, uint32_t aRouterLifetime, RoutePreference aPreference);
```
Comment (jwhui) [inline src/core/net/nd6.hpp:646]:
```suggestion
    Error AppendRouteInfoOption(const Prefix &aPrefix, uint32_t aRouteLifetime, RoutePreference aPreference);
```

File: src/core/border_router/routing_manager.cpp [hunk 5]
```diff
@@ -828,128 +828,76 @@ Error RoutingManager::SendRouterSolicitation(void)
...
         {
-            Ip6::Nd::RouteInfoOption *rio;
-
-            OT_ASSERT(bufferLength +
-                          Ip6::Nd::RouteInfoOption::OptionSizeForPrefix(advertisedOmrPrefix.GetPrefix().GetLength()) <=
-                      sizeof(buffer));
-
-            rio = reinterpret_cast<Ip6::Nd::RouteInfoOption *>(buffer + bufferLength);
+            SuccessOrAssert(
+                raMsg.AppendRouteInfoOption(omrPrefix.GetPrefix(), /* aRouterLifetime */ 0, omrPrefix.GetPreference()));
```
Comment (jwhui) [inline src/core/border_router/routing_manager.cpp:869]:
```suggestion
                raMsg.AppendRouteInfoOption(omrPrefix.GetPrefix(), /* aRouteLifetime */ 0, omrPrefix.GetPreference()));
```

File: src/core/net/nd6.cpp [hunk 6]
```diff
@@ -178,17 +193,81 @@ void RouterAdvMessage::SetToDefault(void)
...
+    pio->SetValidLifetime(aValidLifetime);
+    pio->SetPreferredLifetime(aPreferredLifetime);
+    pio->SetPrefix(aPrefix);
+
+exit:
+    return error;
+}
+
+Error RouterAdvertMessage::AppendRouteInfoOption(const Prefix &  aPrefix,
+                                                 uint32_t        aRouterLifetime,
```
Comment (jwhui) [inline src/core/net/nd6.cpp:253]:
```suggestion
                                                 uint32_t        aRouteLifetime,
```

File: src/core/net/nd6.hpp [hunk 7]
```diff
@@ -445,101 +455,219 @@ class RouteInfoOption : public Option, private Clearable<RouteInfoOption>
...
 
     /**
-     * This method gets the default router preference.
+     * This method indicates whether or not the RA message contains any options.
      *
-     * @returns  The router preference.
+     * @retval TRUE   If the RA message contains at least one option.
+     * @retval FALSE  If the RA message contains no option.
```
Comment (jwhui) [inline src/core/net/nd6.hpp:652]:
```suggestion
     * @retval FALSE  If the RA message contains no options.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**